File: Search a 2D Matrix/Search a 2D Matrix.java
public class Solution {
    public boolean searchMatrix(int[][] matrix, int target) {
        int rowMin = 0;
        int rowMax = matrix.length;
        int rowMid = (rowMin + rowMax) / 2;
        int prevMid = rowMid;
        
        do{
            prevMid = rowMid;
            if(matrix[rowMid][0] == target){
                return true;
            }
            else if(matrix[rowMid][0] < target){
                rowMin = rowMid;
                rowMid = (rowMin + rowMax) / 2;
            }
            else{
                rowMax = rowMid;
                rowMid = (rowMin + rowMax) / 2;
            }
        }while(prevMid != rowMid);
        
        int colMin = 0;
        int colMax = matrix[0].length;
        int colMid = (colMin + colMax) / 2;
        int prevCol = colMid;
        
        do{
            prevCol = colMid;
            if(matrix[rowMid][colMid] == target){
                return true;
            }
            else if(matrix[rowMid][colMid] < target){
                colMin = colMid;
                colMid = (colMin + colMax) / 2;
            }
            else{
                colMax = colMid;
                colMid = (colMin + colMax) / 2;
            }
        }while(prevCol != colMid);
        
        return false;
        
    }
}

File: Search a 2D Matrix/Search a 2D Matrix.cpp
class Solution {
public:

    int findRow(std::vector<std::vector<int> >& matrix, int target, int left, int right){
        if(left == right)   return left;
        
        int mid = (left + right) / 2;
        if((left == mid) || (right == mid)){
            return (matrix[right][0] > target) ? left : right;
        }
        else{
            return (matrix[mid][0] > target) ? findRow(matrix, target, left, mid - 1) : findRow(matrix, target, mid, right);
        }
    }

    int findColomn(std::vector<std::vector<int> >& matrix, int target, int row, int left, int right){
        if(left == right)   return left;
        
        int mid = (left + right) / 2;
        if((left == mid) || (right == mid)){
            return (matrix[row][right] > target) ? left : right;
        }
        else{
            return (matrix[row][mid] > target) ? findColomn(matrix, target, row, left, mid - 1) : findColomn(matrix, target, row, mid, right);
        }
    }

    bool searchMatrix(vector<vector<int> > &matrix, int target) {
        if(matrix.empty())      return false;
        if(matrix[0].empty())   return false;
        
        int rowSize = matrix.size();
        int colomnSize = matrix[0].size();
        int row = findRow(matrix, target, 0, rowSize - 1);
        int colomn = findColomn(matrix, target, row, 0, colomnSize - 1);
        
        if(matrix[row][colomn] == target)   return true;
        else                                return false;
    }
};
